File: go/tests/shard_kv_codec_test.go
// shard_kv_codec_test.go
//
// Verifies ShardKvCodec (ported from ShardKvCodec.cs):
//   - K-path wire bytes match the ground-truth fixture (float32 scale LE +
//     int8 quantised components), the load-bearing cross-language format.
//   - Encode→Decode round-trips K approximately (quantisation is lossy) and V
//     exactly to the selected codeword.
//   - VQ picks the nearest codeword; the returned V equals that codeword.
//   - The V index byte-width follows the codebook size (1/2/4 bytes).
//   - Construction + dim guards reject bad shapes.

package circleai_test

import (
	"bytes"
	"encoding/hex"
	"math"
	"testing"

	circleai "github.com/bhengubv/CircleAI/go"
)

type shardEncodeFixture struct {
	Cases []struct {
		ID          string    `json:"id"`
		KDim        int       `json:"kDim"`
		KRank       int       `json:"kRank"`
		VDim        int       `json:"vDim"`
		VCodewords  int       `json:"vCodewords"`
		K           []float32 `json:"k"`
		Scale       float32   `json:"scale"`
		Projected   []float32 `json:"projected"`
		EncodedKHex string    `json:"encodedK_hex"`
		EncodedVLen int       `json:"encodedV_len"`
	} `json:"cases"`
}

// identityAxes builds the kRank×kDim identity-top-rank axes the ShardKvCodec
// ctor seeds, so the test pins bytes against a known, C#-matching basis.
func identityAxes(kRank, kDim int) [][]float32 {
	axes := make([][]float32, kRank)
	for r := 0; r < kRank; r++ {
		axes[r] = make([]float32, kDim)
		axes[r][r] = 1
	}
	return axes
}

func TestShardKvCodec_EncodeKBytes_Fixture(t *testing.T) {
	var fix shardEncodeFixture
	readLocalFixture(t, "shard_kv_encode.json", &fix)
	if len(fix.Cases) == 0 {
		t.Fatal("no shard cases")
	}
	for _, c := range fix.Cases {
		c := c
		t.Run(c.ID, func(t *testing.T) {
			codec, err := circleai.NewShardKvCodec(c.KDim, c.KRank, c.VDim, c.VCodewords, 0)
			if err != nil {
				t.Fatalf("ctor: %v", err)
			}
			// Axes = identity-top-rank (the ctor already seeds this, but set it
			// explicitly so the intent is clear and independent of ctor state).
			if err := codec.SetPrincipalAxes(identityAxes(c.KRank, c.KDim)); err != nil {
				t.Fatalf("set axes: %v", err)
			}
			v := make([]float32, c.VDim) // all-zero V → deterministic nearest codeword

			frame, err := codec.Encode(c.K, v)
			if err != nil {
				t.Fatalf("encode: %v", err)
			}
			want, _ := hex.DecodeString(c.EncodedKHex)
			if !bytes.Equal(frame.CompressedK, want) {
				t.Errorf("CompressedK bytes mismatch:\n got %x\nwant %x", frame.CompressedK, want)
			}
			if len(frame.CompressedV) != c.EncodedVLen {
				t.Errorf("CompressedV length: got %d want %d", len(frame.CompressedV), c.EncodedVLen)
			}
			if frame.KOriginalDim != c.KDim || frame.VOriginalDim != c.VDim {
				t.Errorf("frame dims: got (%d,%d) want (%d,%d)", frame.KOriginalDim, frame.VOriginalDim, c.KDim, c.VDim)
			}
		})
	}
}

func TestShardKvCodec_RoundTrip_KApprox_VExact(t *testing.T) {
	codec, err := circleai.NewShardKvCodec(8, 8, 4, 16, 0)
	if err != nil {
		t.Fatalf("ctor: %v", err)
	}
	// Full-rank identity axes so K reconstruction is exact up to int8 quantisation.
	codec.SetPrincipalAxes(identityAxes(8, 8))

	// Inject a known codebook so V decode is deterministic and independent of
	// any RNG. Row i is the constant vector (i/10).
	cb := make([][]float32, 16)
	for i := range cb {
		cb[i] = []float32{float32(i) / 10, float32(i) / 10, float32(i) / 10, float32(i) / 10}
	}
	if err := codec.SetVCodebook(cb); err != nil {
		t.Fatalf("set codebook: %v", err)
	}

	k := []float32{0.9, -0.4, 0.2, 0.05, -0.7, 0.33, 0.1, -0.15}
	// V closest to codeword 5 (=0.5 everywhere).
	v := []float32{0.51, 0.49, 0.5, 0.52}

	frame, err := codec.Encode(k, v)
	if err != nil {
		t.Fatalf("encode: %v", err)
	}
	gotK, gotV, err := codec.Decode(frame)
	if err != nil {
		t.Fatalf("decode: %v", err)
	}

	// K: lossy — require close reconstruction (int8 over an 8-dim vector).
	for i := range k {
		if d := math.Abs(float64(gotK[i] - k[i])); d > 0.03 {
			t.Errorf("K[%d] reconstruction off: got %v want ~%v (|d|=%v)", i, gotK[i], k[i], d)
		}
	}
	// V: exact codeword copy — codeword 5.
	for i := range gotV {
		if math.Abs(float64(gotV[i]-0.5)) > 1e-6 {
			t.Errorf("V[%d]: got %v want 0.5 (nearest codeword)", i, gotV[i])
		}
	}
}

func TestShardKvCodec_VQ_PicksNearestCodeword(t *testing.T) {
	codec, err := circleai.NewShardKvCodec(4, 4, 3, 4, 0)
	if err != nil {
		t.Fatalf("ctor: %v", err)
	}
	cb := [][]float32{
		{0, 0, 0},
		{1, 1, 1},
		{-1, -1, -1},
		{5, 5, 5},
	}
	codec.SetVCodebook(cb)

	// Query nearest codeword 1 = (1,1,1).
	frame, err := codec.Encode([]float32{0, 0, 0, 0}, []float32{0.9, 1.1, 1.0})
	if err != nil {
		t.Fatalf("encode: %v", err)
	}
	if got := int(frame.CompressedV[0]); got != 1 {
		t.Errorf("expected codeword index 1, got %d", got)
	}
}

func TestShardKvCodec_VIndexByteWidth(t *testing.T) {
	// 256 codewords → 1 byte; 512 → 2 bytes. (4-byte path needs >65536 which is
	// impractical to allocate here; the 1/2 boundary exercises the switch.)
	cases := []struct {
		codewords int
		wantBytes int
	}{
		{codewords: 256, wantBytes: 1},
		{codewords: 512, wantBytes: 2},
	}
	for _, c := range cases {
		codec, err := circleai.NewShardKvCodec(4, 4, 2, c.codewords, 0)
		if err != nil {
			t.Fatalf("ctor codewords=%d: %v", c.codewords, err)
		}
		frame, err := codec.Encode([]float32{0, 0, 0, 0}, []float32{0, 0})
		if err != nil {
			t.Fatalf("encode codewords=%d: %v", c.codewords, err)
		}
		if len(frame.CompressedV) != c.wantBytes {
			t.Errorf("codewords=%d: CompressedV len got %d want %d", c.codewords, len(frame.CompressedV), c.wantBytes)
		}
	}
}

func TestShardKvCodec_ObserveK_UpdatesMean(t *testing.T) {
	codec, _ := circleai.NewShardKvCodec(3, 3, 2, 4, 0)
	if codec.SamplesObserved() != 0 {
		t.Fatalf("fresh codec should have 0 samples")
	}
	codec.ObserveK([]float32{2, 4, 6})
	codec.ObserveK([]float32{4, 8, 12})
	if codec.SamplesObserved() != 2 {
		t.Errorf("expected 2 samples, got %d", codec.SamplesObserved())
	}
	// Mean of the two samples is (3,6,9); with a non-zero centre, encoding a
	// vector equal to the mean projects (after Hadamard) to ~0, so the scale
	// floors at ~1e-9 and all int8 components are 0.
	codec.SetPrincipalAxes(identityAxes(3, 3))
	frame, err := codec.Encode([]float32{3, 6, 9}, []float32{0, 0})
	if err != nil {
		t.Fatalf("encode: %v", err)
	}
	for i := 4; i < len(frame.CompressedK); i++ {
		if frame.CompressedK[i] != 0 {
			t.Errorf("component %d should be 0 when input equals mean, got %d", i-4, int8(frame.CompressedK[i]))
		}
	}
}

func TestShardKvCodec_ConstructionGuards(t *testing.T) {
	if _, err := circleai.NewShardKvCodec(0, 1, 4, 4, 0); err == nil {
		t.Error("kDim=0 should error")
	}
	if _, err := circleai.NewShardKvCodec(4, 5, 4, 4, 0); err == nil {
		t.Error("kRank>kDim should error")
	}
	if _, err := circleai.NewShardKvCodec(4, 2, 0, 4, 0); err == nil {
		t.Error("vDim=0 should error")
	}
	if _, err := circleai.NewShardKvCodec(4, 2, 4, 3, 0); err == nil {
		t.Error("non-power-of-two codewords should error")
	}
	if _, err := circleai.NewShardKvCodec(4, 2, 4, 1, 0); err == nil {
		t.Error("codewords=1 should error")
	}
}

func TestShardKvCodec_DimMismatchGuards(t *testing.T) {
	codec, _ := circleai.NewShardKvCodec(4, 2, 3, 4, 0)
	if _, err := codec.Encode([]float32{1, 2, 3}, []float32{0, 0, 0}); err == nil {
		t.Error("wrong K dim should error")
	}
	if _, err := codec.Encode([]float32{1, 2, 3, 4}, []float32{0, 0}); err == nil {
		t.Error("wrong V dim should error")
	}
	if err := codec.ObserveK([]float32{1, 2}); err == nil {
		t.Error("ObserveK wrong dim should error")
	}
	if err := codec.SetPrincipalAxes([][]float32{{1, 0, 0, 0}}); err == nil {
		t.Error("wrong axes shape should error")
	}
	if err := codec.SetVCodebook([][]float32{{0, 0, 0}}); err == nil {
		t.Error("wrong codebook size should error")
	}
}

File: rust/tests/shard_kv_codec_test.rs
//! shard_kv_codec_test.rs
//!
//! Verifies [`circle_ai::model_runtime::ShardKvCodec`] — the byte-exact port of
//! `CircleAI.Core.Compression.ShardKvCodec`. Checks construction guards, the wire
//! format (float32-LE scale header + int8 body, LE codeword index), the
//! deterministic seed codebook, round-trip fidelity, and V exact reconstruction.

use circle_ai::model_runtime::{ShardCompressedFrame, ShardKvCodec};

#[test]
fn rejects_bad_dimensions() {
    assert!(ShardKvCodec::new(0, 1, 4, 4, 0).is_err());
    // kRank > kDim.
    assert!(ShardKvCodec::new(8, 9, 4, 4, 0).is_err());
    // vDim == 0.
    assert!(ShardKvCodec::new(8, 4, 0, 4, 0).is_err());
    // vCodewords not a power of two.
    assert!(ShardKvCodec::new(8, 4, 4, 3, 0).is_err());
    // vCodewords == 1.
    assert!(ShardKvCodec::new(8, 4, 4, 1, 0).is_err());
    // Valid.
    assert!(ShardKvCodec::new(8, 4, 4, 4, 0).is_ok());
}

#[test]
fn encoded_k_wire_format_is_scale_header_plus_int8_body() {
    let mut codec = ShardKvCodec::new(8, 4, 4, 4, 0).unwrap();
    let k = [0.5f32, -0.25, 0.75, 0.1, -0.6, 0.2, 0.9, -0.3];
    let v = [0.1f32, 0.2, 0.3, 0.4];
    let frame = codec.encode(&k, &v).unwrap();

    // encoded_k = 4-byte float32 LE scale + kRank int8 bytes.
    assert_eq!(frame.compressed_k.len(), 4 + 4);
    let mut scale_bytes = [0u8; 4];
    scale_bytes.copy_from_slice(&frame.compressed_k[0..4]);
    let scale = f32::from_le_bytes(scale_bytes);
    assert!(scale > 0.0, "scale should be positive, got {scale}");

    // Frame carries flattened axes of length kRank * kDim.
    assert_eq!(frame.k_principal_axes.len(), 4 * 8);
    assert_eq!(frame.k_original_dim, 8);
    assert_eq!(frame.v_original_dim, 4);
}

#[test]
fn v_index_byte_width_tracks_codebook_size() {
    // 4 codewords → 1 byte index.
    let mut small = ShardKvCodec::new(4, 2, 3, 4, 0).unwrap();
    let f = small.encode(&[0.1, 0.2, 0.3, 0.4], &[0.5, 0.6, 0.7]).unwrap();
    assert_eq!(f.compressed_v.len(), 1);

    // 512 codewords → 2-byte index.
    let mut big = ShardKvCodec::new(4, 2, 3, 512, 0).unwrap();
    let f2 = big.encode(&[0.1, 0.2, 0.3, 0.4], &[0.5, 0.6, 0.7]).unwrap();
    assert_eq!(f2.compressed_v.len(), 2);
}

#[test]
fn v_decodes_to_exact_codeword() {
    // With identity setup, V is VQ'd to the nearest codeword and decode returns
    // that codeword exactly (copy, not lossy).
    let mut codec = ShardKvCodec::new(4, 2, 4, 8, 7).unwrap();
    let k = [1.0f32, 0.0, 0.0, 0.0];
    let v = [0.9f32, -0.9, 0.1, 0.5];
    let frame = codec.encode(&k, &v).unwrap();
    let (_k_out, v_out) = codec.decode(&frame).unwrap();

    // v_out must equal one of the codebook words — specifically the nearest.
    // We can at least assert it round-trips stably: re-encoding v_out picks the
    // same index and decodes to the identical vector.
    let frame2 = codec.encode(&v_out, &v_out).unwrap();
    let (_k2, v_out2) = codec.decode(&frame2).unwrap();
    assert_eq!(v_out, v_out2);
}

#[test]
fn k_round_trips_within_quantisation_error() {
    // No observed samples → centre is zero. Identity-top-rank axes + self-inverse
    // Hadamard means K decode reconstructs the original up to int8 error.
    let k_dim = 8;
    let mut codec = ShardKvCodec::new(k_dim, k_dim, 4, 4, 0).unwrap();
    let k: Vec<f32> = (0..k_dim).map(|i| (i as f32 - 4.0) * 0.1).collect();
    let v = [0.1f32, 0.2, 0.3, 0.4];

    let frame = codec.encode(&k, &v).unwrap();
    let (k_out, _v_out) = codec.decode(&frame).unwrap();

    assert_eq!(k_out.len(), k_dim);
    // int8 quantisation over 8 dims — generous but bounded error.
    for (orig, got) in k.iter().zip(k_out.iter()) {
        assert!(
            (orig - got).abs() < 0.05,
            "K reconstruction drift too large: orig={orig}, got={got}"
        );
    }
}

#[test]
fn observe_k_updates_running_mean_count() {
    let mut codec = ShardKvCodec::new(4, 2, 4, 4, 0).unwrap();
    assert_eq!(codec.samples_observed(), 0);
    codec.observe_k(&[1.0, 2.0, 3.0, 4.0]).unwrap();
    codec.observe_k(&[5.0, 6.0, 7.0, 8.0]).unwrap();
    assert_eq!(codec.samples_observed(), 2);
    // Wrong dim rejected.
    assert!(codec.observe_k(&[1.0, 2.0]).is_err());
}

#[test]
fn set_principal_axes_shape_is_validated() {
    let mut codec = ShardKvCodec::new(4, 2, 4, 4, 0).unwrap();
    // Correct shape (2 x 4).
    let axes = vec![vec![1.0, 0.0, 0.0, 0.0], vec![0.0, 1.0, 0.0, 0.0]];
    assert!(codec.set_principal_axes(&axes).is_ok());
    // Wrong shape.
    let bad = vec![vec![1.0, 0.0, 0.0]];
    assert!(codec.set_principal_axes(&bad).is_err());
}

#[test]
fn set_v_codebook_is_validated() {
    let mut codec = ShardKvCodec::new(4, 2, 3, 4, 0).unwrap();
    let cb = vec![
        vec![0.1, 0.2, 0.3],
        vec![0.4, 0.5, 0.6],
        vec![0.7, 0.8, 0.9],
        vec![1.0, 1.1, 1.2],
    ];
    assert!(codec.set_v_codebook(&cb).is_ok());
    // Wrong count.
    assert!(codec.set_v_codebook(&cb[..3].to_vec()).is_err());
    // Wrong dim.
    let bad = vec![
        vec![0.1, 0.2],
        vec![0.4, 0.5],
        vec![0.7, 0.8],
        vec![1.0, 1.1],
    ];
    assert!(codec.set_v_codebook(&bad).is_err());
}

#[test]
fn seed_codebook_is_deterministic() {
    // Two codecs with the same seed pick identical codewords, so encoding the
    // same V yields identical compressed_v.
    let mut a = ShardKvCodec::new(4, 2, 4, 16, 123).unwrap();
    let mut b = ShardKvCodec::new(4, 2, 4, 16, 123).unwrap();
    let k = [0.3f32, 0.1, -0.2, 0.5];
    let v = [0.6f32, -0.4, 0.2, 0.8];
    let fa = a.encode(&k, &v).unwrap();
    let fb = b.encode(&k, &v).unwrap();
    assert_eq!(fa.compressed_v, fb.compressed_v);
    assert_eq!(fa.compressed_k, fb.compressed_k);
}

#[test]
fn decode_rejects_dim_mismatched_frame() {
    let mut codec = ShardKvCodec::new(4, 2, 4, 4, 0).unwrap();
    let bad_frame = ShardCompressedFrame::new(vec![0u8; 6], vec![0u8; 1], vec![0.0; 8], 8, 4);
    assert!(codec.decode(&bad_frame).is_err());
}
